### ISSUE TITLE
Directionality for ROIs

### ISSUE THREAD

**Issue by bancroftg** [sgoldenlab/simba#396]
Unable to determine directionality (in respect to ROIs):

 When I go to "Aggregate Boolean Conditional Statistics" I do not have any conditional rule behavior options "facing" the ROIs, only options where the animal is "in zone". This is after successfully drawing ROIs, analyzing ROI data aggregates, and extracting features. After extracting features, I appended ROI features by body part.

When I initially created this project, I did not use a SimBA suggested animal body-part config, but instead produced my own. It is 4 points with a center point instead of the tail. Am I required to use a SimBA body-part config in order to compute directionality?

Please let me know and thank you!

**Comment by sronilsson** [comment]
Hi @bancroftg!

In order to compute where the animal is facing in relation to ROIs and other animals in SimBA GUI, it requires body-parts for the left ear, nose, and right ear. It's done like the image below.

However, if SimBA cannot find a body part name that contains the two words `left` and `ear`, and another body part containing the two words `right` and `ear`, and a third body with the word `nose`, it doesn't know which body-part is which and it won't be direction/facing will not be computed.

**What are the names of your 4 body-parts you are tracking and what body-parts are they?** (PS. [its possible to use the nape of the neck](https://simba-uw-tf-dev.readthedocs.io/en/latest/simba.circular_statistics.html#simba.mixins.circular_statistics.CircularStatisticsMixin.direction_two_bps) as well if you don't have the ears tracked but it would have to be done programmatically and not in GUI. 











![image](https://github.com/user-attachments/assets/c23e6f97-99a6-423c-a511-357c6774e254)

**Comment by bancroftg** [comment]
I have left ear, right ear, nose, and center. When I append the features do they need to be in any specific order?

**Comment by sronilsson** [comment]
No, no specific order required - can you send me an example CSV file from the `project_folder/csv/features_extracted` directory? 

Do you know if you have columns in those files with the `_facing` suffix?

**Comment by bancroftg** [comment]
[5218_dSLR_test_A1.csv](https://github.com/user-attachments/files/17318310/5218_dSLR_test_A1.csv)

Not seeing _facing suffux

**Comment by sronilsson** [comment]
Cheers, let me check the code to remind myself what could be happening.

**Comment by sronilsson** [comment]
Ah yeah I can see..  you use a space in the bodypart name, while the code tries to split the bodypart name exlusively by `_`.... so 'left_ear` or `ear_left` works while `ear left` or `left ear` doesn't work. Let me improve that code.

**Comment by bancroftg** [comment]
Thank you!

**Comment by sronilsson** [comment]
I inserted a fix that also works on your body-parts as you named them. Just got to take the dog for her walk lol, I will push the update when I get back and let you know.

**Comment by sronilsson** [comment]
If you update Simba with `pip install simba-uw-tf-dev` upgrade, how does it look on your end if you try to run the ROI feature extraction again followed by the conditional booleans?

**Comment by bancroftg** [comment]
I ran the Simba update as `pip install simba-uw-tf-dev --upgrade` and ran feature extraction again. This gave me the directionality info perfectly! Thank you!

**Comment by sronilsson** [comment]
@bancroftg  cool - one question: I just ran this boolean conditional analysis, and it just outputted a single aggregate count of when the behaviors happened (or not happened) at the same time. I didn't get any values for the exact time period bouts when they happened at the same time. Do you think that would be useful to add?

**Comment by bancroftg** [comment]
Oh Definitely! And the features extracted csv's give a frame-by-frame analysis, correct?

**Comment by bancroftg** [comment]
I have another small issue actually: When I now go to "Visualize ROI features" with the following settings, I'm running into an error:
![image](https://github.com/user-attachments/assets/5b26c18d-6d11-4b76-b161-16d675181c6d)
![image](https://github.com/user-attachments/assets/5bd8e531-e496-46e1-ba88-02f9f8eb416b)

**Comment by sronilsson** [comment]
Ah yeah, let me fix that too - while I do:  how does it look if you tick the `Multi-process` checkbox, doe sit run then?

**Comment by Enricocloud** [comment]
Hi, sorry if I popped up in this conversation, but my issues might be related to this topic.
I want to calculate directionality. I already used these features before the updates, but I archived my old data and added new tracking data from Sleap. 

But the GUI does not allow me to select the option now.
I could not understand why from the guide.

Any idea?

Thank you 

![Image](https://github.com/user-attachments/assets/22286b8e-f783-4d06-b3ec-44433dbdf026)

**Comment by sronilsson** [comment]
Hi @Enricocloud - I do yeah. 

I inserted a check a little earlier, to see if the nose, left ear, right was tracked for the animal. If that is not the case, I decided to grey this option out. 

Thanks for the screenshot super helpful. 

To determine if you have nose, left ear, right tracked in the project, SimBA looks at your project body-part names (stored in `project_folder\logs\measures\pose_configs\bp_names\project_bp_names.csv`), for different combinations of the word `right` (or `left`) paired with `ear` (and a bosy-part with the text `nose` in it. In your case the body-part is named `R_ear` (and I presume `L_ear`) and the code will not recognize that as the right ear and left ear.

Is it possible that you named your body-part `right_ear` or `ear_right` or similarly previously when it worked?

If you rename the body-parts `right_ear` or `ear_right`  etc,  in your project and data, doe sit become un-greyed?

**Comment by Enricocloud** [comment]
AH.
Yes, now it is possible to use the function.

I renamed the node in the project_bp_names.csv with right_ear and left_ear.

Thank you very much

**Comment by Heinrich2818** [comment]
> Hi [@bancroftg](https://github.com/bancroftg)!
> 
> In order to compute where the animal is facing in relation to ROIs and other animals in SimBA GUI, it requires body-parts for the left ear, nose, and right ear. It's done like the image below.
> 
> However, if SimBA cannot find a body part name that contains the two words `left` and `ear`, and another body part containing the two words `right` and `ear`, and a third body-part with the word `nose`, it doesn't know which body-part is which and direction/facing will not be computed.
> 
> **What are the names of your 4 body-parts you are tracking and what body-parts are they?** (PS. [its possible to use the nape of the neck](https://simba-uw-tf-dev.readthedocs.io/en/latest/simba.circular_statistics.html#simba.mixins.circular_statistics.CircularStatisticsMixin.direction_two_bps) as well if you don't have the ears tracked but it would have to be done programmatically and not in GUI.
> 
> ![image](https://github.com/user-attachments/assets/c23e6f97-99a6-423c-a511-357c6774e254)

Hi @sronilsson !
I’d also like to analyze my animals’ directionality, but my tracking data uses body part names like "earL" and "earR". This means I can’t use the directionality analysis directly in SimBA at the moment.

Would it be possible to add an option in the GUI that allows users to manually select which three body parts in their track data (e.g., left ear, right ear, and nose/earL" , "earR" and "nose") should be used to compute directionality?

That would make the feature much more flexible for users with different naming conventions. 🙏

**Comment by sronilsson** [comment]
Yes, that's an excellent point @Heinrich2818.. not sure why I didn't think of that earlier lol 🙃

Which GUI menu/button are you using to compute this, do you have a screenshot? Helps me jog my memory where to insert this.

Thanks
Simon

**Comment by Heinrich2818** [comment]
> Yes, that's an excellent point [@Heinrich2818](https://github.com/Heinrich2818).. not sure why I didn't think of that earlier lol 🙃
> 
> Which GUI menu/button are you using to compute this, do you have a screenshot? Helps me jog my memory where to insert this.
> 
> Thanks Simon

Of course @sronilsson !
I'm computing directionality in following menu:

<img width="979" height="626" alt="Image" src="https://github.com/user-attachments/assets/17c9aff3-e27d-481a-8a89-ee9f0cff41aa" />

<img width="524" height="59" alt="Image" src="https://github.com/user-attachments/assets/b1a99a6b-7f01-4776-9110-d5835e3a9ea6" />

<img width="685" height="676" alt="Image" src="https://github.com/user-attachments/assets/bfaf613f-a52c-40c9-bc89-30a7e87c2714" />

**Comment by sronilsson** [comment]
Thanks @Heinrich2818 - I will insert drop downs to specify the body part names. I am travelling at the moment though, so will likely be a little slower than usual - but will let you know when done.

**Comment by sronilsson** [comment]
hey @Heinrich2818 - when you click to analyze directionality between animals, you should now see 3 drop-downs where you can specify the names for your three body-parts. Update to latest version of simba with `pip install simba-uw-tf-dev -U` to see them. These three drop-downs will be pre-populated with the body-part names in your project closest matching `ear_left`, `ear_right`, `nose` as per SimBAs best guess. Change them as needed: 

<img width="1136" height="593" alt="Image" src="https://github.com/user-attachments/assets/49096162-e93a-4337-bb9f-0f9133e321d1" />

When visualizing directionality, you will also have the same, three, drop-down options:

<img width="1335" height="520" alt="Image" src="https://github.com/user-attachments/assets/94774530-893f-40fd-b8a4-89e9265bfc65" />


Can you please let me know how it runs on your end? Also, if you have seen it, could you point me to the part of the documentation where these methods are described? I want to update the documentation so others don't get confused by these new drop-downs.

**Comment by Heinrich2818** [comment]
> hey [@Heinrich2818](https://github.com/Heinrich2818) - when you click to analyze directionality between animals, you should now see 3 drop-downs where you can specify the names for your three body-parts. Update to latest version of simba with `pip install simba-uw-tf-dev -U` to see them. These three drop-downs will be pre-populated with the body-part names in your project closest matching `ear_left`, `ear_right`, `nose` as per SimBAs best guess. Change them as needed:
> 
> <img alt="Image" width="1136" height="593" src="https://private-user-images.githubusercontent.com/34761092/510288307-49096162-e93a-4337-bb9f-0f9133e321d1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjMwODMyNzgsIm5iZiI6MTc2MzA4Mjk3OCwicGF0aCI6Ii8zNDc2MTA5Mi81MTAyODgzMDctNDkwOTYxNjItZTkzYS00MzM3LWJiOWYtMGY5MTMzZTMyMWQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTE0VDAxMTYxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2YzJlYjE2MTdhNzMwZjU2ZjI5YjkyNDgwMDEzYzMzOThjM2Q0MDk5MmRiMTA2NGJmOTNlOTUwZTc3ZWEyYTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xKUd3YMrHPKCZKlALBVdoBnKUYfluJrEvqkxcotkXdg">
> When visualizing directionality, you will also have the same, three, drop-down options:
> 
> <img alt="Image" width="1335" height="520" src="https://private-user-images.githubusercontent.com/34761092/510289668-94774530-893f-40fd-b8a4-89e9265bfc65.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjMwODMyNzgsIm5iZiI6MTc2MzA4Mjk3OCwicGF0aCI6Ii8zNDc2MTA5Mi81MTAyODk2NjgtOTQ3NzQ1MzAtODkzZi00MGZkLWI4YTQtODllOTI2NWJmYzY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTE0VDAxMTYxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5ZTZhYWY4YzE1ZDc5NzEwZWMwM2FjY2FkY2EyMmM5NzI1NGMwNDBiY2UxMGFiNTE0YzA3NzJlN2U5NDI2MGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZaBz9rHea4fp1RVKIgb8KEnC_W7sF2snk27OAxhnXik">
> Can you please let me know how it runs on your end? Also, if you have seen it, could you point me to the part of the documentation where these methods are described? I want to update the documentation so others don't get confused by these new drop-downs.

Sorry for late reply @sronilsson ,I’ve been busy with some presentations and forms recently and didn’t have time to test it until today.

I checked your patch — it works perfectly at startup with the default settings, and I can successfully generate the directionality report:

<img width="652" height="412" alt="Image" src="https://github.com/user-attachments/assets/4efc3ec6-6490-4829-bffe-3140e9314d0c" />

<img width="1100" height="107" alt="Image" src="https://github.com/user-attachments/assets/ad81f69f-bd73-463f-9354-d21faab568cf" />

Then I tested whether I could manually select bps to calculate directionality:

<img width="661" height="412" alt="Image" src="https://github.com/user-attachments/assets/16428ed5-f133-4d82-98b3-e293ca406cf6" />

I noticed that every bp had a postfix like  '_1'. When I manually select a bp, I can no longer use the default bp names (without postfix) to calculate directionality. Using a manually selected bp leads to an error:

<img width="605" height="367" alt="Image" src="https://github.com/user-attachments/assets/fd80650e-0ab8-450c-908e-b0a9cb0262ee" />

<img width="1121" height="92" alt="Image" src="https://github.com/user-attachments/assets/a5b626f3-c7aa-4a66-885a-6d98e62a16c1" />

**Comment by sronilsson** [comment]
Ah thanks @Heinrich2818 - I appreciate the testing. Let me look over this and see what's going on with those suffixes...

**Comment by sronilsson** [comment]
I think I mistakenly listed the body-part names **with** suffixes as options in the dropdowns, but did the right thing and pre-populated the dropdown **without** the suffixes.  

Can you see if it runs in the latest version of simba 4.2.1?

**Comment by Heinrich2818** [comment]
> I think I mistakenly listed the body-part names **with** suffixes as options in the dropdowns, but did the right thing and pre-populated the dropdown **without** the suffixes.  
> 
> Can you see if it runs in the latest version of simba 4.2.1?  

I updated to the latest version, and it works perfectly now! Thanks for your great work @sronilsson

**Comment by sronilsson** [comment]
Thanks @Heinrich2818 ! If you see any more issues, including with the directionality  visualization function if you use it, please let me know !
